src/Interpreters/executeQuery.cpp

```diff
@@ -699,10 +700,52 @@ static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
                 if (OpenTelemetry::CurrentContext().isTraceEnabled())
                 {
                     auto * raw_interpreter_ptr = interpreter.get();
-                    std::string class_name(demangle(typeid(*raw_interpreter_ptr).name()));
+                    String class_name(demangle(typeid(*raw_interpreter_ptr).name()));
                     span = std::make_unique<OpenTelemetry::SpanHolder>(class_name + "::execute()");
                 }
+
                 res = interpreter->execute();
+
+                /// Try to read (SELECT) query result from query result cache (if it is enabled)
+                auto query_result_cache = context->getQueryResultCache();
+                bool read_result_from_query_result_cache = false; /// a query must not read from *and* write to the query result cache at the same time
+                if (query_result_cache != nullptr
+                    && (settings.enable_experimental_query_result_cache || settings.enable_experimental_query_result_cache_passive_usage)
+                    && res.pipeline.pulling())
+                {
+                    QueryResultCache::Key key(
+                        ast, settings.query_result_cache_partition_key, res.pipeline.getHeader(),
+                        std::make_optional<String>(context->getUserName()),
+                        std::chrono::system_clock::now() + std::chrono::seconds(settings.query_result_cache_ttl));
+                    QueryResultCache::Reader reader = query_result_cache->createReader(key);
+                    if (reader.hasCacheEntryForKey())
+                    {
+                        res.pipeline = QueryPipeline(reader.getPipe());
```

Comment:
Yes, if a cache entry exists, the original pipeline is returned in `explain pipeline`. The advantage of that is that the internal implementation is not exposed. But tbh, one could equally argue for showing the replaced pipeline because this is what actually runs.

Anyways, the reason for the behavior is a bit silly:
- in query `SELECT ... SETTINGS enable_experimental_query_result_cache = true`, the SETTINGS clause applies to `SELECT` (obviously)
- in query `EXPLAIN (PLAN|SETTINGS) SELECT ... SETTINGS enable_experimental_query_result_cache = true` (*) the `SETTINGS` clause applies to `EXPLAIN`. When the inner `SELECT` is processed, `enable_experimental_query_result_cache = false` and the QRC is bypassed 🤦 

I checked that for query (*) no entry is created in the QRC. It isn't but because I don't really understand the EXPLAIN machinery yet, I can't explain yet why. Guess I should add a test to make sure the behavior doesn't break unexpectedly.